On the "My competitions" page show competitions without a date or country

pingskills opened the pull request:
Fixes #495. I modified competitions_controller.rb to put all competitions without a start date first and then sort all competitions that have a start date in reverse order. To ensure the date_range was displayed without error I modified the wca_date_range method to only return the date_range if both a from_date and to_date are passed in. In the view I fixed a bug with <tbody> being shown for each row. Only display the country.name if the country has been selected.

Let me know your thoughts.

pingskills commented:
The second commit fixes #415. It turns out that a regular competitor could not register for a non visible competition since the registrations_controller.rb has the following code:
`    if !competition.user_can_view?(current_user)
      raise ActionController::RoutingError.new('Not Found')
    end`

Organisers could register though so I updated the code to stop this from occurring. Let me know what you think.

jonatanklosko commented on WcaOnRails/spec/controllers/registrations_controller_spec.rb:
```diff
@@ -161,6 +161,15 @@
+
+    it "cannot accept own registration for a competition that isn't visible" do
+      competition.showAtAll = 0
+      competition.save!
```
Use `competition.update_column(:showAtAll, false)` =)

pedrosino replied:
Maybe update! so it raises an exception?

jonatanklosko replied:
Do we need to run the validations? I think we can just update the table directly. We have something similar [here](https://github.com/cubing/worldcubeassociation.org/blob/master/WcaOnRails/spec/controllers/competitions_controller_spec.rb#L90-L96).

pedrosino replied:
You're probably right...I don't know for sure when to raise exceptions or not :P

larspetrus replied:
I think you _always_ run the validations, unless you have a real good reason not to!

Even if skipping validation is fine with the current code, that can change, silently turning this code into a bug.

jfly replied:
I agree with Lars, we should always run validations unless there's a good reason not to. I'm ok with taking the performance hit in out tests.

jonatanklosko replied:
Got it, this is reasonable.

jonatanklosko commented on WcaOnRails/spec/controllers/registrations_controller_spec.rb:
```diff
@@ -222,6 +231,15 @@
 
+    it "cannot create registration when competition is not visible" do
+      competition.showAtAll = 0
+      competition.save!
```
Same as above.

larspetrus commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -313,7 +313,7 @@ def my_competitions
-    @competitions = @competitions.uniq.sort_by(&:start_date).reverse!
+    @competitions = @competitions.uniq.reject(&:start_date) + @competitions.uniq.select(&:start_date).sort_by(&:start_date).reverse!
     @not_past_competitions = @competitions.reject(&:is_over?)
     @past_competitions = @competitions.select(&:is_over?)
```
@past_competitions, @not_past_competitions = @competitions. partition(&:is_over?)

pingskills replied:
Nice :)

larspetrus commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -313,7 +313,7 @@ def my_competitions
       @competitions += current_user.person.competitions
     end
-    @competitions = @competitions.uniq.sort_by(&:start_date).reverse!
+    @competitions = @competitions.uniq.reject(&:start_date) + @competitions.uniq.select(&:start_date).sort_by(&:start_date).reverse!
```
Something like this should work:

`@competitions = @competitions.uniq.sort_by { |comp| comp.start_date || Date.new(2100, 1, 1) }.reverse`

I tried getting rid of the `reverse`, but it's weirdly hard to sort date in reverse order.

jfly replied:
I like @larspetrus's suggestion, but let's use `Date::Infinity` instead?

pingskills replied:
Nice, I was going to use:

@competitions = @competitions.uniq.sort_by { |comp| comp.start_date || Date.today + 20.year }.reverse

pingskills replied:
Looks like I can't compare a Date with Date:Infinity ?

`@competitions = @competitions.uniq.sort_by { |comp| comp.start_date || Date::Infinity.new }.reverse`

Error message: comparison of Date::Infinity with Date failed

jfly replied:
Wow, that is super lame. How about taking advantage of lexicographical sorting?

```
@competitions = @competitions.uniq.sort_by { |comp| [ comp.start_date ? 0 : 1, comp.start_date ] }.reverse
```

jfly replied:
Gah, that probably won't work either. The `... + 20.year` sounds okay to me =(

larspetrus commented on WcaOnRails/app/controllers/competitions_controller.rb:
```diff
@@ -313,7 +313,7 @@ def my_competitions
     if current_user.person
```
It's not your fault that `my_competitions` doesn't have a test, but something this complex really needs one!

jfly replied:
I agree with @larspetrus, this is worthy of a test.

jfly commented on WcaOnRails/app/views/competitions/_my_competitions_table.html.erb:
```diff
@@ -18,6 +18,7 @@
       </tr>
     </thead>
+    <tbody>
     <% competitions.each do |competition| %>
```
Indentation?

jfly commented on WcaOnRails/app/views/registrations/register.html.erb:
```diff
@@ -42,7 +42,7 @@
   <% else %>
 
-    <% if @competition.registration_opened? %>
+    <% if @competition.registration_opened? && @competition.showAtAll? %>
```
This removes the form, but does this actually prevent organizers from registering for a competition whose registration is open, but `showAtAll` is false?

jfly replied:
I personally think it's good that we allow organizers to register for an invisible competition, although it is understandable that it confused @Luis-J-Ianez. How about we just add an alert that makes it very clear that they're allowed to register for their comp, but other people even with the link won't be able to?

Also, can you create a separate PR for this? I'd rather not have different issues being address in the same PR.

pingskills replied:
I had added some code in the controller to stop the organiser from registering. I get what you are saying though and had some similar thoughts myself in that an organiser should be able to register. As the competition warnings are already displayed on the page there is now a warning that says "This competition is not visible to the public". So I think I can just add a test showing that a user will get a 'Not Found' exception, and that an organiser can register? I'll put this together soon in a new PR so people can comment.

pingskills commented:
I've pulled out some changes so that this PR is only related to #495. I'll put the other changes into a separate PR to keep things clean as suggested by @jfly.

Let me know your thoughts.

jfly commented on WcaOnRails/app/helpers/application_helper.rb:
```diff
@@ -124,6 +124,8 @@ def process_nav_items(nav_items)
   def wca_date_range(from_date, to_date, options={})
     options[:separator] = '-'
-    date_range(from_date, to_date, options)
+    if from_date && to_date
```
I'd rather this method blow up if `from_date` or `to_date` is nil, and force the people calling this method to handle that case explicitly.

pingskills replied:
Cool. I'll revert this change and handle it in the view.

jfly replied:
Great, while you're at it, could you audit the other callers and see if they're checking as well?

pingskills replied:
Will do.

pingskills commented:
Reset wca_date_range method so it doesn't perform any checks. Modified the 3 calls of this method where a check didn't already exist to perform a check and display "Unscheduled" if the comp doesn't have start and end dates.

As for tests for my_competitions, does anyone have any suggestions or examples of similar tests?

pingskills commented:
I've added some tests for my_competitions. Let me know if you think they are appropriate and if there are other tests that would be useful.

jonatanklosko commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -571,4 +571,96 @@
+      let!(:competition4) { FactoryGirl.create(:competition, :registration_open, starts: 2.month.ago, organizers: [organizer1], eventSpecs: "222 333") }
+      let!(:waldner_user) { FactoryGirl.create :user, name: "Jan-Ove Waldner" }
+      let!(:registration1) { FactoryGirl.create(:registration, competitionId: competition1.id, user: waldner_user) }
+      let!(:registration2) { FactoryGirl.create(:registration, competitionId: competition3.id, user: delegate1) }
```
Move all these `let!`s from all your tests level up, directly under the describe line. There's no reason to repeat them.

jonatanklosko commented:
Referring to the line @larspetrus commented before, I think there should be a test for the person's (that's associated with a user) competitions.

jonatanklosko commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -571,4 +571,96 @@
+      let!(:competition2) { FactoryGirl.create(:competition, :registration_open, starts: 1.month.ago, organizers: [organizer1], delegates: [delegate1], eventSpecs: "222 333") }
+      let!(:competition3) { FactoryGirl.create(:competition, :registration_open, starts: 2.weeks.from_now, organizers: [organizer1], eventSpecs: "222 333") }
+      let!(:competition4) { FactoryGirl.create(:competition, :registration_open, starts: 3.weeks.from_now, organizers: [organizer1], eventSpecs: "222 333") }
+      let!(:competition4) { FactoryGirl.create(:competition, :registration_open, starts: 2.month.ago, organizers: [organizer1], eventSpecs: "222 333") }
```
Nitpick: the competition is past so the registration shouldn't be open =)

pingskills replied:
True, I thought I had to have the registration open though to add a registration. And I wanted to show past_competitions and not_past_competitions. Is there another way to add the registration to a past competition?

jonatanklosko replied:
Oh, I see. That's ok then.

jonatanklosko commented on WcaOnRails/app/views/competitions/_index_table.html.erb:
```diff
@@ -17,7 +17,7 @@
           <%= icon('hourglass-start', title: "Starts in #{ pluralize((competition.start_date - Date.today).to_i, "day") }", data: { toggle: "tooltip" }) %>
         <% end %>
-        <%= wca_date_range(competition.start_date, competition.end_date)  %>
+        <%= competition.start_date && competition.end_date ? wca_date_range(competition.start_date, competition.end_date) : "Unscheduled" %>
```
We definitely don't want such a `"Unscheduled"` competitions to be shown on the competitions index. I think they aren't placed in the `@competitions`.

pingskills replied:
That is true, unscheduled competitions shouldn't end up in @competitions but if they did it would throw an error. I thought it was better to put this in as a safeguard. Happy to remove it if you think it's not required.

jonatanklosko replied:
Related to this. Just found out a funny issue. If you create a competition it hasn't a start and end dates (which is the correct behavior) but you can make it visible immediately. One of the results is that on the index page you can select `Past from 0` year (there's no exception so it's filtered out in the controller probably by chance). @jfly is it because we rely on people that they won't make such a incomplete competition visible?
So I think we don't need this `"Unscheduled"`. If there's a really danger that such a comp may became visible, then we should rather prevent from this directly.

jfly replied:
Oh yikes. We have a number of additional validations for confirming a competition (including requiring that [dates be present](https://github.com/jfly/worldcubeassociation.org/blob/9998edd118e34b0d86df316ebb2a58ffab6f64fe/WcaOnRails/app/models/competition.rb#L396-L400)), but there's probably nothing that would prevent a board member from making a competition and then making it visible without first confirming it. We should enforce this "extra validations" both when confirming and when making a competition visible to the public.

@pingskills, in the interest of not repeating ourselves, I've changed my mind and decided I'd be ok with moving the "Unscheduled" out of the views and into the `wca_date_range` method. In my previous job, we had a very strong culture of writing code that would fail-fast, rather than limping along in some scenario the programmer didn't intend for. My knee jerk reaction is to apply that philosophy to everything, but I think the conciseness and DRY-ness of moving the "Unscheduled" string into the `wca_date_range` method is worth it here. Also, it would be *really* awesome if you pulled the "Unscheduled" string out into a yml file, as per @viroulep's recent work to start getting the website translated (see https://github.com/cubing/worldcubeassociation.org/pull/460).

jfly replied:
I've created https://github.com/cubing/worldcubeassociation.org/issues/604 to track the additional Competition validations. @pingskills, please don't worry about taking care of that in this PR.

pingskills replied:
Itis really good that the website is being translated! When moving "Unscheduled" into the en.yml do I also need to update fr.yml? What happens if there is an English version but no French version? Google translate tells me that unscheduled is imprévu in French :)

jfly replied:
Don't worry about keeping the French version up to date. Rails will fall back to the English version of it can't find a translation.

pingskills commented:
I cleaned up the tests based on feedback from @jonatanklosko

jonatanklosko commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -571,4 +571,79 @@
+    let!(:future_competition3) { FactoryGirl.create(:competition, :registration_open, starts: 1.weeks.from_now, organizers: [organizer1], eventSpecs: "222 333") }
+    let!(:past_competition4) { FactoryGirl.create(:competition, :registration_open, starts: 1.month.ago, organizers: [organizer1], eventSpecs: "222 333") }
+    let!(:past_competition5) { FactoryGirl.create(:competition, :registration_open, starts: 2.month.ago, organizers: [organizer2], delegates: [delegate1], eventSpecs: "222 333") }
+    let!(:past_competition6) { FactoryGirl.create(:competition, :registration_open, starts: 3.month.ago, organizers: [organizer2], delegates: [delegate1], eventSpecs: "222 333") }
```
Nitpick: `past_competition4-6` -> `1-3`

pingskills replied:
When I updated the name to past_competition I thought to myself I should change the numbers too, but then I was too lazy :)

jonatanklosko commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -571,4 +571,79 @@
+    let!(:past_competition4) { FactoryGirl.create(:competition, :registration_open, starts: 1.month.ago, organizers: [organizer1], eventSpecs: "222 333") }
+    let!(:past_competition5) { FactoryGirl.create(:competition, :registration_open, starts: 2.month.ago, organizers: [organizer2], delegates: [delegate1], eventSpecs: "222 333") }
+    let!(:past_competition6) { FactoryGirl.create(:competition, :registration_open, starts: 3.month.ago, organizers: [organizer2], delegates: [delegate1], eventSpecs: "222 333") }
+    let!(:waldner_user) { FactoryGirl.create :user, name: "Jan-Ove Waldner" }
```
Maybe just `user`? The name here isn't necessary and it would be simpler =)

jfly commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -571,4 +571,79 @@
+      end
+    end
+
+    context 'when signed in as a regular user' do
```
This looks great, thanks for writing the test!

I know this is already pretty intense, but can you add a test for competitions that a user competed in, but didn't register for? This can happen for competitions whose registration wasn't done on the WCA website, for instance.

pingskills replied:
In fact I'm an example of that. Here in Australia we don't use the WCA website for registrations.

larspetrus commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -571,4 +571,79 @@
     end
   end
+
+  describe 'GET #my_competitions' do
```
This is a good and ambitious test! 

Some improvements/suggestions:

* I don't see any test of the ` || Date.today + 20.year` part. If you remove that, I assume the test still passes? You probably need a competition without start date.

* The value of @competitions is not tested. If it's used, it should be tested. If not, it should lose the @.

* The "one assertion per test" pattern gets very wordy and a bit confusing here. I would do them like this:

    context 'when signed in as an organizer user' do
      it 'shows my past and upcoming competitions' do
        sign_in organizer1

        get :my_competitions

        expect(assigns(:not_past_competitions)).to eq [future_competition1, future_competition2, future_competition3]
        expect(assigns(:past_competitions)).to eq [past_competition1, past_competition3]
      end
    end

pingskills replied:
Thanks @larspetrus. I'll create a test without a start_date and clean up the tests as you suggested.

pingskills replied:
@competitions is not used - I'll change it to competitions.

pingskills commented:
I've tried to clean up the existing tests. I added a new test where the start date is not set and another new test to ensure that a user who has not registered but has results still sees the competition.

I updated the wca_date_range method to return "unscheduled" if either date is not passed in. The last thing I want to do is put unscheduled into a yml file.

Let me know your thoughts :)

jonatanklosko commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -571,4 +571,101 @@
+          average: 2000,
+          regionalSingleRecord: "WR",
+          regionalAverageRecord: "WR",
+        )
```
Actually, why wouldn't we use
```ruby
FactoryGirl.create(:result, person: results_person, competitionId: competition1.id)
```

pingskills replied:
Good idea, I thought I tried something similar but must have had a typo somewhere. I'll update this.

larspetrus commented on WcaOnRails/spec/controllers/competitions_controller_spec.rb:
```diff
@@ -571,4 +571,101 @@
+      end
+
+      it 'shows my upcoming and past competitions' do
+        unscheduled_competition1.update_columns(year: 0, month: 0, day: 0, endMonth: 0, endDay: 0)
```
This looks good, but why not set these columns when FactoryGirl creates competition on line 584?

pingskills replied:
I tried to set those in line 584 but couldn't get it to work. When I used starts: "" or starts: nil I get an error and if I try to set the year, month, and day it doesn't seem to actually set them? If someone could help me out here with the right syntax that would be appreciated.

pingskills replied:
Ahh, maybe I need to update spec/factories/competitions.rb to allow an empty string?

jfly replied:
Ahhh, yeah, it sounds like it needs to be able to handle nil.

jonatanklosko replied:
`starts` and `ends` are just a FactoryGirl transient attributes, which allow us to set the competition dates in a handy way and they have to be the Date objects. If you want these dates to be nil you can just set `start_date: nil` directy, and `end_date` in the same way.

pingskills replied:
Hi @jonatanklosko but then in the factory the following code is called:
    start_date { starts.strftime("%F") }
    end_date { ends.strftime("%F") }
which causes an error when I pass in `start_date: nil` so I think I need to update those 2 lines to handle being passed a nil?
```

jonatanklosko replied:
Did you check it? I was sure that if we specify the atttibute, the block doesn't execute

pingskills replied:
Yes I checked it and it failed.

jonatanklosko replied:
And did you try to set both start and end dates to nil in this way? If you specify only the start_date, the validation throws an exception because we compare the dates unless both are nil (I'm not able to check it right now and that's why I cannot play with it).

pingskills replied:
Yeah, I tried putting starts:nil, ends: nil but got the same error regarding the date. I've added a test to check if they are nil to the factory and it seems to work well now.

jonatanklosko replied:
Ahh, I meant `start_date: nil` and `end_date: nil`. It should set the attributes directly, so we don't use `starts` and `ends` in this case.

jfly replied:
Yeah, what jonatan says should be true. That said, I think handling starts and ends is a nice step towards following the principle of least surprise :)

pingskills commented:
OK. I have updated the competition factory so we can directly pass a nil start and end date, then cleaned up the rest of the tests.

Still outstanding is to move the "unscheduled" string into the en.yml file.

pingskills commented:
I've now put the string unscheduled into the en.yml file.

I could be wrong but I think this is getting close to being finished :)

timhabermaas commented:
68 comments, wow. I applaud your endurance, @pingskills 👍 

There's one thing I'd like to mention though: Rails 5 will remove controller tests, especially `assigns`, see https://github.com/rails/rails/issues/18950 for the reasons (tl;dr tests are brittle). The translation of this test to a test which tests the entire stack wouldn't be too difficult, so 👍 for merging from me. We should probably stop testing against instance variables in the future though.

jfly commented:
LGTM!

jfly commented:
Merged and pushed to production. https://www.worldcubeassociation.org/competitions/mine seems to still be working for me, woohoo!

pingskills commented:
Excellent. Thanks for your help everyone :)
